### PR TITLE
Remove smoke point maximum

### LINKED ISSUE #20805: Smoke point maximum doesn't work
### Type

Game behaviour

### Bug description

This slipped by when the code was simplified to rely on function calls instead of event subscription (https://github.com/ppy/osu/pull/20376/commits/6628ab5190409e93dd4fb46f57ac0877c1d27905). Before, `SmokeSegment` stopped adding points by unsubscribing, but when that was replaced with function calls, there was no more unsubscription. So even though `FinishDrawing` is called when the max point count is hit, that function doesn't actually do anything to stop `SmokeSegment` from adding more points.

---

I was gonna open PR a fix, but I saw two ways about it and wasn't sure which was better:
1. Add an `isActive` variable which is set to `false` in `FinishDrawing`, and early return if `isActive` is `false`
2. Move all of the max point count logic to `SmokeContainer` and use the pre-existing `currentSegmentSkinnable` variable

Method 1 is easier/simpler, but I was leaning towards Method 2 since it separates responsibility more clearly, where `SmokeContainer` sorta "manages" `SmokeSegment` (spawning + input handling, and now capping too).

### Screenshots or videos

_No response_

### Version

2022.1008.2

### Logs

[legacy-ipc.log](https://github.com/ppy/osu/files/9805009/legacy-ipc.log)
[network.log](https://github.com/ppy/osu/files/9805010/network.log)
[performance.log](https://github.com/ppy/osu/files/9805011/performance.log)
[performance-draw.log](https://github.com/ppy/osu/files/9805012/performance-draw.log)
[performance-update.log](https://github.com/ppy/osu/files/9805013/performance-update.log)
[runtime.log](https://github.com/ppy/osu/files/9805014/runtime.log)
[updater.log](https://github.com/ppy/osu/files/9805015/updater.log)
[database.log](https://github.com/ppy/osu/files/9805016/database.log)
[input.log](https://github.com/ppy/osu/files/9805017/input.log)

### PR DESCRIPTION
**Pull request by goodtrailer**
Closes #20805.

As pointed out, stable smoke is uncapped, so lazer's should be too. This requires some behavior to be rewritten/optimized. It's still not super performant, but it does perform better than stable on my machine (using the same replay where I had max sensitivity and just shook my mouse wildly in circles for a solid minute and a half). Very not rigorous benchmarking though.

Major changes:
1. Keep track of the whole smoke segment, even future points (if rewinding), and then only copy the relevant part of the smoke segment to the `SmokeDrawNode`
2. Make quad batch expand dynamically

Minor optimizations:
1. Early return from `SmokeDrawNode.drawPointQuad` to avoid scale/rotation calculations
2. Reorder `SmokeDrawNode.PointColour` to only calculate initial fade out when necessary (the diff/comparison gets very scuffed)
3. Change `SmokePoint` to hold an angle instead of a direction, which avoids an atan2 in `SmokeDrawNode.PointDirection`

### CODE REVIEW

File: osu.Game.Rulesets.Osu/Skinning/SmokeSegment.cs [hunk 1]
```diff
@@ -234,9 +236,14 @@ public sealed override void Draw(IRenderer renderer)
...
                     return;
 
-                rotationIndex = 0;
+                quadBatch ??= renderer.CreateQuadBatch<TexturedVertex2D>(200, 4);
+
+                if (points.Count > quadBatch.Size * 4 && quadBatch.Size != 10922)
+                {
+                    int batchSize = Math.Min((int)(quadBatch.Size * 1.5f), 10922);
+                    quadBatch = renderer.CreateQuadBatch<TexturedVertex2D>(batchSize, 4);
+                }
```
Comment (smoogipoo) [inline osu.Game.Rulesets.Osu/Skinning/SmokeSegment.cs:245]: - `VertexBatch<T>.Size` actually indicates the number of elements (in this case, quads), contrary to the documentation which states number of vertices. So if each point is adding 4 vertices (= `IRenderer.VERTICES_PER_QUAD`) then I'd expect the first part of the condition to be `if (points.Count > quadBatch.Size)`.
- `10922` can be replaced with `IRenderer.MAX_QUADS`.
- Any reason for the 1.5 expansion factor? List implementations typically [scale with a factor of 2](https://source.dot.net/#System.Private.CoreLib/src/libraries/System.Private.CoreLib/src/System/Collections/Generic/List.cs,456) which sounds fine here.

**Review by smoogipoo — CHANGES_REQUESTED** [review]
Other than that, this looks fine to me

File: osu.Game.Rulesets.Osu/Skinning/SmokeSegment.cs (hunk 1 above)
Comment (goodtrailer) [reply osu.Game.Rulesets.Osu/Skinning/SmokeSegment.cs:245]: I think some C++ `std::vector` implementations use 1.5x, but 2x seems equally common based on a quick search. If .NET `Lists` use 2x then we can go with 2x, I didn't really have any particular reason for 1.5x. (I see some references to 1.5x being closer to the golden ratio, but I honestly have no idea how that's relevant)

**Review by smoogipoo — APPROVED** [review]
looks good / feels good to me

**Review by peppy — APPROVED** [review]

**Pull request merged**